analyses/comparative-RNASeq-analysis/01-correlation-matrix.py

```diff
@@ -0,0 +1,107 @@
+
+https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/229
+"""
+
+import os
+import sys
+import argparse
+import math
+import numpy as np
+import pandas as pd
+import scipy.stats
+import sklearn.metrics.pairwise as sklp
+import utils
+
+def prepare_expression(input_matrix,
+    proportion_unexpressed=0.8,
+    variance_filter_level=0.2,
+    verbose=False,
+    prefix=""):
+    """Loads samples, converts to log2(TPM+1), and applies expression and variance filters.
+    returns dataframe of converted and filtered expression.
+    writes file: scratch/{prefix}filtered_genes_to_keep.rds"""
+
+    print_v = print if verbose else lambda *a, **k: None
+    filtered_genelist_filename = "{}filtered_genes_to_keep.rds".format(prefix)
+
+    print_v("Loading samples {}".format(input_matrix))
+    raw_samples = utils.read_rds(input_matrix, location="data")
+
+    # Convert to log2(tpm+1)
+    samples = raw_samples.apply(lambda x: np.log2(x+1))
+
+    # run expression filter
+    # Remove any genes that have 0 expression in more samples than proportion_unexpressed
+    print_v("Running expression filter")
+    max_ok_zeroes = len(samples.columns) * proportion_unexpressed
+    has_few_enough_zeroes = samples.apply(
+        lambda s: len(s[s<=0]) < max_ok_zeroes,
+        axis=1)
+    expression_filtered_df = samples[has_few_enough_zeroes]
+
+    # run variance filter
+    # Sort remaining genes by variance and drop the least-varying
+    # How many to drop is controlled by variance_filter_level
+    print_v("Running variance filter")
+    variance = expression_filtered_df.apply(np.std, axis=1)
+    cut_proportion = int(math.ceil(len(variance)*variance_filter_level))
+    keep_proportion = len(variance) - cut_proportion
+
+    expr_var_filtered_genelist = pd.DataFrame(variance.nlargest(keep_proportion).index)
+    expr_var_filtered_df = expression_filtered_df.filter(expr_var_filtered_genelist["gene_id"], axis=0)
+
+    utils.write_rds(expr_var_filtered_genelist, filtered_genelist_filename, location="scratch")
+
+    return expr_var_filtered_df
+
+
+def calculate_correlation(expr_var_filtered_df, verbose=False, prefix=""):
+    print_v = print if verbose else lambda *a, **k: None
+    """Takes an expression matrix dataframe and calculates spearman correlations.
+    (spearman is a rank transformed pearson ('correlation') metric)
+    output file: scratch/{prefix}all_by_all_correlations.tsv"""
+
+    all_by_all_filename = "{}all_by_all_correlations.rds".format(prefix)
+
+    # Rank transform - values are now that rank of that gene within each sample
+    print_v("Rank transforming")
+    filtered_rank_transformed_df = np.apply_along_axis(scipy.stats.rankdata,0,expr_var_filtered_df)
+
+    # Run pearson metric on ranked genes. For pairwise_distances we need samples=rows so transpose.
+    print_v("Running pairwise distances")
+    x_corr = 1 - sklp.pairwise_distances(X=filtered_rank_transformed_df.transpose(), metric="correlation")
```

Comment:
I'm curious here if there is a reason not to use `scipy.stats.spearmanr()`? I would think it should have the same result, so it shouldn't really matter, but I'm curious if there might be some issue that this method addresses that led you to use it here?

Reply:
Yeah, I also expect the result should be the same. This part of the code is copied pretty directly from our [UCSC TumorMap library]( https://github.com/Stuartlab-UCSC/hexmap-data/blob/master/calc/compute_sparse_matrix.py#L332) so that we can potentially generate TumorMaps from these data.